source/extensions/common/wasm/context.cc

```diff
@@ -1234,8 +1238,11 @@ WasmResult Context::defineMetric(uint32_t metric_type, std::string_view name,
     return WasmResult::BadArgument;
   }
   auto type = static_cast<MetricType>(metric_type);
+  // Prefix the given name with CustomStatNamespacePrefix so that these user-defined
+  // custom metrics can be distinguished from native Envoy metrics.
+  const auto prefixed_name = absl::StrCat(CustomStatNamespacePrefix, std::string(name));
   // TODO: Consider rethinking the scoping policy as it does not help in this case.
-  Stats::StatNameManagedStorage storage(toAbslStringView(name), wasm()->scope_->symbolTable());
+  Stats::StatNameManagedStorage storage(prefixed_name, wasm()->scope_->symbolTable());
```

Comment:
I think this is a topic for another PR, but I think we should talk about having the wasm API be more symbol-table friendly. This is not a new topic -- I was trying to influence the API design earlier when other people were working on WASM, but I am certain this is going to be a perf bottleneck if this is called by every worker thread in an extension.

For now, would you do 2 things:
1. Leave behind a TODO to revisit this
2. Make an incremental improvement to this code now by storing the symbolized CustomStatNamespacePrefix (without the ".") in the Context object and using `Stats::Utility::counterFromStatNames` (etc) below? That would at least hold the global symbol table lock for less time as it would factor out the tokenization and symbolization of your static prefix.

Reply:
Thanks, that sounds great. I added a TODO comment: https://github.com/envoyproxy/envoy/pull/17357/commits/b6dc70d8da4f355229d9dae529033476089ed0a2. will follow-up in another PR soon!

Reply:
Sorry, my suggestion was a more aggressive on both points:

1. I think you can go ahead and store the symbolized prefix in the context this PR (this is a very incremental change to the code you are editing) and use counterFromStatNames.

2. In a TODO let's address (in a broader forum) the concern that users of wasm will take global symbol table locks every time they create a new couter, even if the names they are composing for the counter are known at compile time. This is will be real performance bottleneck on servers with lots of cores and threads, and I think the API needs to be changed. I brought this up much earlier during WASM's development and I think it's a real problem. That's beyond the scope of this PR and it'd be great to leave a TODO for that. @PiotrSikora may have some context on this.

Reply:
OK just resolved 1. in thanks, deleted in https://github.com/envoyproxy/envoy/pull/17357/commits/fcf4035a992e4cfb30c5b1728991057b6301df59. As for 2, I haven't left that as a TODO - since the API redesign itself does not fit in one TODO in context.cc. Also I believe as long as the users of Wasm extensions uses "defineMetric" API only in the initialization (i.e. only during the startup phase of Wasm VM), this wouldn't be that severe, although I agree we should consider some alternatives like posting definMetric API calls into main thread, etc.

Reply:
If the wasm user guide recommends a pattern of declaring stats when an extension loads, and before it goes multi-threaded, then performance should be OK when that recommendation is followed.

However, if an extension-writer creates a stat in a worker thread from a string in wasm, performance will be poor. This is why I think the API should guide the developer toward creating symbolic stat names for any stat they might potentially want to create in a worker thread. Because that definitely happens in C++ extensions. For example there's a large but bounded set of known cipher names in SSL that are pre-symbolized at startup, but we don't create actual counters for those until they are used.

I don't have sign-off on the wasm spec, so my main conduit for emphasizing this is PR reviews. How do we get this resolved? @PiotrSikora ?

Reply:
@jmarantz if you believe that this is a real concern, then please report it using the proper channel (i.e. open a new issue or a PR with the fix), so that it can be discussed there. Side-discussion in the comments of another PR is a recipe for the issue to be lost.